src/Microsoft.AspNetCore.Routing/Matchers/TreeMatcher.cs

```diff
@@ -0,0 +1,348 @@
+                }
+            }
+
+            var trees = new List<UrlMatchingTree>();
+            for (var i = 0; i < entries.Count; i++)
+            {
+                var entry = entries[i];
+
+                while (trees.Count <= entry.Order)
+                {
+                    trees.Add(new UrlMatchingTree(entry.Order));
+                }
+
+                var tree = trees[entry.Order];
+                tree.AddEntry(entry);
+            }
+
+            return trees.ToArray();
+        }
+
+        private InboundRouteEntry MapInbound(RouteTemplate template, Endpoint[] endpoints, int order)
+        {
+            if (template == null)
+            {
+                throw new ArgumentNullException(nameof(template));
+            }
+
+            var entry = new InboundRouteEntry()
+            {
+                Precedence = RoutePrecedence.ComputeInbound(template),
+                RouteTemplate = template,
+                Order = order,
+                Tag = endpoints,
+            };
+
+            var constraintBuilder = new RouteConstraintBuilder(_constraintFactory, template.TemplateText);
+            foreach (var parameter in template.Parameters)
+            {
+                if (parameter.InlineConstraints != null)
+                {
+                    if (parameter.IsOptional)
+                    {
+                        constraintBuilder.SetOptional(parameter.Name);
+                    }
+
+                    foreach (var constraint in parameter.InlineConstraints)
+                    {
+                        constraintBuilder.AddResolvedConstraint(parameter.Name, constraint.Constraint);
+                    }
+                }
+            }
+
+            entry.Constraints = constraintBuilder.Build();
+
+            entry.Defaults = new RouteValueDictionary();
+            foreach (var parameter in entry.RouteTemplate.Parameters)
+            {
+                if (parameter.DefaultValue != null)
+                {
+                    entry.Defaults.Add(parameter.Name, parameter.DefaultValue);
+                }
+            }
+            return entry;
+        }
+
+        private struct Key : IEquatable<Key>
+        {
+            public readonly int Order;
+            public readonly string Template;
+
+            public Key(int order, string routePattern)
+            {
+                Order = order;
+                Template = routePattern;
+            }
+
+            public bool Equals(Key other)
+            {
+                return Order == other.Order && string.Equals(Template, other.Template, StringComparison.OrdinalIgnoreCase);
+            }
+
+            public override bool Equals(object obj)
+            {
+                return obj is Key ? Equals((Key)obj) : false;
+            }
+
+            public override int GetHashCode()
+            {
+                var hash = new HashCodeCombiner();
```

Comment:
First time I've seen someone use this 👍

Reply:
Oh, it's all over MVC and Razor. 

They are adding something like this to the BCL I think???